#Where's Schueller?

Professor Schueller allows his phone to track his location resulting in google location data from 2013-2024. This dataset holds the key to answering questions about Schueller's routines, preferences and habits. What's Schueller's preferred grocery store? Where does he spend the majority of his time on campus? What time does he arrive to school? All these questions can be answered using this data.

## Data Information

This dataset is a pickle file of data that tracked gps location of professor Schueller from the years 2013-2024. It includes the longitude and latitude of Schueller's locations and its respective timestamp/datetime. The variables of note are therefore: `latitude`, `longitude`, `TimeStamp` (which will be used as a hover heading) and `datetime` which is the index of the dataset.

In [ ]:
# import the necessary libraries
import pandas as pd
import pickle
import numpy as np
from datetime import datetime
import plotly.express as px
from geopy.distance import great_circle

In [ ]:
# load the pickle file
file1 = '/content/drive/MyDrive/GPS2.pkl'

df = pd.read_pickle(file1)

#Grocery Stores

There are five prominent supermarket chains in Walla Walla Washington. These include Safeway on Rose street, Safeway on Plaza Way, Albertsons, Grocery Outlet and Super 1 Foods. Of these which does Professor Schueller prefer to shop at? Safeway is the closest to campus, however grocery outlet often has great bargains, but maybe Super 1 Foods is the only store to sell a specific good. To find out, we first obtain coordinates for each store from google maps. Next, new columns containing the distance between a coordinate to the respective grocery store will be added to the data frame.

* Note: Walmart is not included here because it is not considered a grocery store and is instead labeled a "superstore." Additionally, the Walmart is located in College Place while the rest of the stores are in Walla Walla.

In [ ]:
# Lat/Lon of multiple grocery stores in town
safeway_rose = (46.07020167488385, -118.33782571841355)
safeway_plaza = (46.048652540313924, -118.34805732019333)
albertsons = (46.07664316195397, -118.30284129146864)
grocery_outlet = (46.05408139839594, -118.34682127478426)
super_foods = (46.05581902452754, -118.3475630001305)

# Create new columns for the distance from each point to each location
df['dist_rose'] = df.apply(lambda x: great_circle((x['latitude'], x['longitude']), safeway_rose).meters, axis=1)
df['dist_plaza'] = df.apply(lambda x: great_circle((x['latitude'], x['longitude']), safeway_plaza).meters, axis=1)
df['dist_albertsons'] = df.apply(lambda x: great_circle((x['latitude'], x['longitude']), albertsons).meters, axis=1)
df['dist_grocery_outlet'] = df.apply(lambda x: great_circle((x['latitude'], x['longitude']), grocery_outlet).meters, axis=1)
df['dist_super_foods'] = df.apply(lambda x: great_circle((x['latitude'], x['longitude']), super_foods).meters, axis=1)

Next, new data frames for each store are be created from the original, which include only locations within 20 meters of the respective grocery store coordinates.

In [ ]:
# Create dataframes for which the point is within 20 meters of each location
df_rose = df[df['dist_rose'] <= 20]
df_plaza = df[df['dist_plaza'] <= 20 ]
df_albertsons = df[df['dist_albertsons'] <= 20]
df_grocery_outlet = df[df['dist_grocery_outlet'] <= 20]
df_super_foods = df[df['dist_super_foods'] <= 20]

Perhaps we could be done here, we can count how many points are within 20 meters of a store and determine Shueller's shopping preferences from the number of points in each dataframe.

In [ ]:
count_rose1 = len(df_rose)
count_plaza1 = len(df_plaza)
count_albertsons1 = len(df_albertsons)
count_grocery_outlet1 = len(df_grocery_outlet)
count_super_foods1 = len(df_super_foods)

print(f"Approximate Visits to Safeway (Rose): {count_rose1}")
print(f"Approximate Visits to Safeway (Plaza): {count_plaza1}")
print(f"Approximate Visits to Albertsons: {count_albertsons1}")
print(f"Approximate Visits to Grocery Outlet: {count_grocery_outlet1}")
print(f"Approximate Visits to Super Foods: {count_super_foods1}")

Approximate Visits to Safeway (Rose): 614
Approximate Visits to Safeway (Plaza): 333
Approximate Visits to Albertsons: 146
Approximate Visits to Grocery Outlet: 15
Approximate Visits to Super Foods: 43


However, `datetime` index of each dataframe does not have equal intervals. When the GPS location data is tracked can be different and is overall not consistent. For example, if Schueller was in Safeway on Rose for 20 minutes google recorded the location every five minutes, we would have 4 data points representing this one visit. If Schueller was in Albertsons for 20 minutes and google recorded the location every four minutes, we would have 5 datapoints representing this one visit. Therefore, the data must be resampled to obtain equal intervals. Here we resampled the data to record the average location every 20 minutes. If there is no data for a specific time period, it is recorded as an NA. Therefore, we must also drop any rows not containing data.

In [ ]:
# Resample time data, recreate time TimeStamp and Datetime, Drop NAs
df_rose_20min = df_rose[['latitude', 'longitude']].resample('20T').mean()
df_rose_20min['TimeStamp'] = df_rose_20min.index.strftime('%Y-%m-%d %H:%M:%S')
df_rose_20min['DateTime'] = df_rose_20min.index
df_rose_20min = df_rose_20min.dropna(subset = ['latitude', 'longitude'])

df_plaza_20min = df_plaza[['latitude', 'longitude']].resample('20T').mean()
df_plaza_20min['TimeStamp'] = df_plaza_20min.index.strftime('%Y-%m-%d %H:%M:%S')
df_plaza_20min['DateTime'] = df_plaza_20min.index
df_plaza_20min = df_plaza_20min.dropna(subset = ['latitude', 'longitude'])

df_albertsons_20min = df_albertsons[['latitude', 'longitude']].resample('20T').mean()
df_albertsons_20min['TimeStamp'] = df_albertsons_20min.index.strftime('%Y-%m-%d %H:%M:%S')
df_albertsons_20min['DateTime'] = df_albertsons_20min.index
df_albertsons_20min = df_albertsons_20min.dropna(subset = ['latitude', 'longitude'])

df_grocery_outlet_20min = df_grocery_outlet[['latitude', 'longitude']].resample('20T').mean()
df_grocery_outlet_20min['TimeStamp'] = df_grocery_outlet_20min.index.strftime('%Y-%m-%d %H:%M:%S')
df_grocery_outlet_20min['DateTime'] = df_grocery_outlet_20min.index
df_grocery_outlet_20min = df_grocery_outlet_20min.dropna(subset = ['latitude', 'longitude'])

df_super_foods_20min = df_super_foods[['latitude', 'longitude']].resample('20T').mean()
df_super_foods_20min['TimeStamp'] = df_super_foods_20min.index.strftime('%Y-%m-%d %H:%M:%S')
df_super_foods_20min['DateTime'] = df_super_foods_20min.index
df_super_foods_20min = df_super_foods_20min.dropna(subset = ['latitude', 'longitude'])

<ipython-input-6-9c5015317904>:2: FutureWarning: 'T' is deprecated and will be removed in a future version, please use 'min' instead.
  df_rose_20min = df_rose[['latitude', 'longitude']].resample('20T').mean()
<ipython-input-6-9c5015317904>:7: FutureWarning: 'T' is deprecated and will be removed in a future version, please use 'min' instead.
  df_plaza_20min = df_plaza[['latitude', 'longitude']].resample('20T').mean()
<ipython-input-6-9c5015317904>:12: FutureWarning: 'T' is deprecated and will be removed in a future version, please use 'min' instead.
  df_albertsons_20min = df_albertsons[['latitude', 'longitude']].resample('20T').mean()
<ipython-input-6-9c5015317904>:17: FutureWarning: 'T' is deprecated and will be removed in a future version, please use 'min' instead.
  df_grocery_outlet_20min = df_grocery_outlet[['latitude', 'longitude']].resample('20T').mean()
<ipython-input-6-9c5015317904>:22: FutureWarning: 'T' is deprecated and will be removed in a future version, please use 'min

We will also add a column called `source` indicating which store the data is coming from. While these are not useful stored in separate dataframes, they will become useful when plotting the location data.

In [ ]:
# add a source column to each dataframe
df_rose_20min['source'] = 'Safeway (Rose)'
df_plaza_20min['source'] = 'Safeway (Plaza)'
df_albertsons_20min['source'] = 'Albertsons'
df_grocery_outlet_20min['source'] = 'Grocery Outlet'
df_super_foods_20min['source'] = 'Super Foods'

However, there is still a problem with our index. What if Shueller spent an hour in the grocery store? Since we indexed by every 20 minutes, this would still show up as 3 separate points in our dataset. To combat this, we will create a mask that calculates the difference in time between a row and the next row's datetime. If the difference is greater than 1 hour the row is kept, if not, it is filtered out. This attempts to limit the number of points in our dataset to the number of separate visits that Schueller made to each store.

In [ ]:
# Create a mask that compares row to its subsequent row
# Check if the time difference > 1 hour
mask_rose = (df_rose_20min['DateTime'].shift(-1) - df_rose_20min['DateTime'] > pd.Timedelta(hours=1)) | df_rose_20min['DateTime'].shift(-1).isna()
mask_plaza = (df_plaza_20min['DateTime'].shift(-1) - df_plaza_20min['DateTime'] > pd.Timedelta(hours=1)) | df_plaza_20min['DateTime'].shift(-1).isna()
mask_albertsons = (df_albertsons_20min['DateTime'].shift(-1) - df_albertsons_20min['DateTime'] > pd.Timedelta(hours=1)) | df_albertsons_20min['DateTime'].shift(-1).isna()
mask_grocery_outlet = (df_grocery_outlet_20min['DateTime'].shift(-1) - df_grocery_outlet_20min['DateTime'] > pd.Timedelta(hours=1)) | df_grocery_outlet_20min['DateTime'].shift(-1).isna()
mask_super_foods = (df_super_foods_20min['DateTime'].shift(-1) - df_super_foods_20min['DateTime'] > pd.Timedelta(hours=1)) | df_super_foods_20min['DateTime'].shift(-1).isna()

# apply mask to dataframes
final_rose = df_rose_20min[mask_rose]
final_plaza = df_plaza_20min[mask_plaza]
final_albertsons = df_albertsons_20min[mask_albertsons]
final_grocery_outlet = df_grocery_outlet_20min[mask_grocery_outlet]
final_super_foods = df_super_foods_20min[mask_super_foods]

### Visualizing the Results

In order to visualize our results, we must combine each of the separate grocery store data frames into one. Next we will plot the location data using a scatter mapbox plot.

In [ ]:
# combine all dataframes into one
combined_df = pd.concat([final_rose, final_plaza, final_albertsons, final_grocery_outlet, final_super_foods], ignore_index=True)

In [ ]:
with open('/content/drive/MyDrive/Senior Year/data science/Tokens/token_HW5.txt') as f:
    mapbox_token = f.read().strip()

In [ ]:
# plot each on a mapbox
fig = px.scatter_mapbox(
    combined_df,
    lat='latitude',
    lon='longitude',
    hover_name='TimeStamp',
    color='source',
    mapbox_style='carto-positron',
    zoom=12,
    title='Scatter Mapbox with For Each Grocery Store'
)

# enable scroll zoom
config = dict({'scrollZoom': True})

fig.update_layout(mapbox_accesstoken=mapbox_token)

fig.show(config = config)

### Results

We can use the same method as above to find the number of points in each data frame.

In [ ]:
count_rose = len(final_rose)
count_plaza = len(final_plaza)
count_albertsons = len(final_albertsons)
count_grocery_outlet = len(final_grocery_outlet)
count_super_foods = len(final_super_foods)

print(f"Approximate Visits to Safeway (Rose): {count_rose}")
print(f"Approximate Visits to Safeway (Plaza): {count_plaza}")
print(f"Approximate Visits to Albertsons: {count_albertsons}")
print(f"Approximate Visits to Grocery Outlet: {count_grocery_outlet}")
print(f"Approximate Visits to Super Foods: {count_super_foods}")

Approximate Visits to Safeway (Rose): 116
Approximate Visits to Safeway (Plaza): 117
Approximate Visits to Albertsons: 26
Approximate Visits to Grocery Outlet: 8
Approximate Visits to Super Foods: 9


According to this outcome, professor Schueller prefers to shop at Safeway over any of the other large grocery stores in town. Over the years Shueller has made approximately 116 visits to Safeway on Rose Street and 117 visits to Safeway on Plaza Way.

#Heat Map

Professor Scheuller is a professor at here at Whitman and is therefore expected to spend the majority of his time in Olin Hall. However, where else does Schueller spend his time on campus? A heat map will show the volume of Schueller's location data on campus.

To achieve this, we first must alter the dataframe to only include points within a 500 meter radius of Whitman College.

In [ ]:
# filter dataframe to include points only within a 500m radius of Whitman
whitman = (46.07201640913145, -118.32966509438066)

df['distance_from_whitman'] = df.apply(lambda x: great_circle((x['latitude'], x['longitude']), whitman).meters, axis=1)

df_whitman = df[df['distance_from_whitman'] <= 500]

The same situation applies here as in the above example: the time intervals are not equal. We must resample the data to find the average location for a specific time interval. This time we will resample using 5 minutes and will still drop the NAs.

Next, we will simply group by the latitude and longitude and count the number of points in each area. Using that count, we can create a density/heat map using mapbox.

In [ ]:
# Resample by 5 minutes
df_whitman_5min = df_whitman[['latitude', 'longitude']].resample('5T').mean()
df_whitman_5min['TimeStamp'] = df_whitman_5min.index.strftime('%Y-%m-%d %H:%M:%S')
df_whitman_5min['DateTime'] = df_whitman_5min.index
df_whitman_5min = df_whitman_5min.dropna(subset = ['latitude', 'longitude'])

<ipython-input-5-08158bebb9f7>:2: FutureWarning: 'T' is deprecated and will be removed in a future version, please use 'min' instead.
  df_whitman_5min = df_whitman[['latitude', 'longitude']].resample('5T').mean()


In [ ]:
# Group by latitude and longitude to count the number of points in each area
point_counts = df_whitman_5min.groupby(['latitude', 'longitude']).size().reset_index(name='count')

In [ ]:
#Create a Heat Map using mapbox
fig = px.density_mapbox(
    point_counts,
    lat='latitude',
    lon='longitude',
    z='count',
    radius=15,
    zoom=15,
    mapbox_style="carto-positron",
    title='Heat Map of Schueller\'s Location at Whitman'
)

# Set the Mapbox access token
fig.update_layout(mapbox_accesstoken=mapbox_token)

config = dict({'scrollZoom': True})

# Show the figure
fig.show(config = config)


In [ ]:
fig.write_html("heatmap.html")

As expected, the heat map illustrates that Shueller spends a large portion of his time on campus in Olin Hall, specifically on the West side. This is the of the building that the Math department operates in. Additionally, Shueller appears to spend quite a bit of time at the BFFC. Other interesting results include Shueller's propensity to be in Sherwood Athletic Center. This could indicate an interest in athletic events such as Volleyball and Basketball. He also appears to spend decent time in Memorial hall, the Science Building, and Reid Campus Center. Shueller is not in Maxey or the Music Building as often. Finally, while not on Whitman Campus, it is included in the 500 meter radius, Shueller appears to enjoy meals at Big House Brew Pub.

# Time to School/Above and Beyond

Can we use this data to figure out what time Schueller arrives to work? In order to do so, we must find a lat/long coordinate that corresponds to the West Side of Olin Hall, where Schueller spends most of his time at work.

In [ ]:
# Set lat/lon of interest (Olin Hall - "Math Side")
Olin = (46.0728205952165, -118.32853712790182)

# Calculate distance from each row to Olin
df['distance'] = df.apply(lambda x: great_circle((x['latitude'], x['longitude']), Olin).meters, axis=1)

Next, we will create a dataframe that only includes points that are within 50 meters of Olin and the point before it is outside of the 50 meter range. This will make it so that only points that are "arriving points" - points that indicate Schueller's arrival to school - are included in the dataframe.

In [ ]:
mask = (df['distance'] <= 50) & (df['distance'].shift(1) > 50)
mask[0] = False

# Filter the DataFrame based on the mask
filtered_df = df[mask]

# Reset the index of the filtered DataFrame
filtered_df = filtered_df.reset_index(drop=True)

# Create a datetime
filtered_df['datetime'] = pd.to_datetime(filtered_df['TimeStamp'])

#make datetime the index
filtered_df = filtered_df.set_index('datetime')

<ipython-input-30-b9ceb22b0979>:2: FutureWarning:

Series.__setitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To set a value by position, use `ser.iloc[pos] = value`



Additionally, we must filter for the school year dates. Each school year starts and ends on different days. Therefore, we will approximate the start and end dates. We will filter the data to only include data from August 25th to May 15th each year.

In [ ]:
#Filter Data for School Year Dates
df_school_year = filtered_df[
    ((filtered_df.index.month == 8) & (filtered_df.index.day >= 25)) |
    ((filtered_df.index.month >= 9) & (filtered_df.index.month <= 12)) |
    ((filtered_df.index.month == 5) & (filtered_df.index.day <= 15)) |
    ((filtered_df.index.month >= 1) & (filtered_df.index.month <= 4))
]

There are some points in our dataset that are in Olin early in the morning. In an attempt to include points in which Schueller arrives to start a new school day, we will filter for the hours after 3am in the morning. We will also filter the data to only include weekdays. If we are interested in when Schueller arrives at school to start his workday, it doesn't make sense to include weekends since those are not official workdays. Then we will find the first recorded time corresponding to each different day. Finally, we will average these times to get Schueller's estimated arrival time.

In [ ]:
#Filter for time
df_time = df_school_year[(df_school_year.index.hour >= 3)]

#Filter for Weekdays
df_weekdays = df_time[df_time.index.dayofweek < 5]

#find the first data point per day
df_first_time = df_weekdays.groupby(df_weekdays.index.date).first()

In [ ]:
# Convert the 'TimeStamp' column to datetime
df_first_time['TimeStamp'] = pd.to_datetime(df_first_time['TimeStamp'])

# Extract the time component and convert to total seconds
df_first_time['time_in_seconds'] = df_first_time['TimeStamp'].dt.hour * 3600 + df_first_time['TimeStamp'].dt.minute * 60 + df_first_time['TimeStamp'].dt.second

# Calculate the average time in seconds
average_seconds = df_first_time['time_in_seconds'].mean()

# Convert average seconds back to hours, minutes, and seconds
avg_hours, remainder = divmod(int(average_seconds), 3600)
avg_minutes, avg_seconds = divmod(remainder, 60)

# Convert to 12-hour format with AM/PM
period = "AM" if avg_hours < 12 else "PM"
avg_hours = avg_hours % 12  # Convert to 12-hour format
avg_hours = 12 if avg_hours == 0 else avg_hours  # Handle midnight case

# Format the average time
average_time_str = f"{avg_hours}:{avg_minutes:02}:{avg_seconds:02} {period}"
print("Average Time:", average_time_str)

Average Time: 3:23:46 PM


Our resulting time is 3:23 pm. This value feels quite incorrect. Most classes end at 4pm at Whitman and therefore it doesn't make sense for Schueller's arrival time to be this late.